### DISCUSSION TITLE
Discussion

### DISCUSSION THREAD

**Discussion by vient** [Wunkolo/qAverageColor#3] [General]
Played with it a bit and have some notes:
* `-mprefer-vector-width` (yeah, again) - setting it to 512 speeds up "serial" version almost 2x on proper AVX-512 machine. Not so serial after all, modern compilers - I use clang 18 - vectorize serial version pretty well
* Different image sizes should be considered? With default 10MP buffer I get 4.5x speedup for AVX512-VNNI version compared with 256-bit serial. Increasing size to 30MP lowers speedup to 3x - did not go further because I have 45MiB of L3 cache on this machine. 1MP gives 6x speedup.
* * With small image sizes may be good to switch from `chrono::high_resolution_clock` to `rdtscp` - if we only need relative speedup, we don't even need to calculate rdtsc frequency.

Overall, I got those results with Intel w7-2495X@4.8GHz, all AVX offsets set to 0, common cmake options `-DCMAKE_CXX_COMPILER=clang-18 -DCMAKE_CXX_FLAGS="-fuse-ld=lld -flto -lstdc++ -lm -Wno-unused-parameter -Wno-deprecated-declarations"`

|Pixels|Serial 256-bit|Serial 512-bit|AVX2|AVX512|AVX512-VNNI|
|-|-|-|-|-|-|
|1MP|1x|1.33x|2.23x|2.18x|2.29x|
|10MP|1x|1.29x|1.86x|2.15x|2x|
|30MP|1x|1.21x|1.57x|1.88x|1.76x|

**Comment by Wunkolo** [comment]
> -mprefer-vector-width

This would be nice to compare the optimized versions against to make things a little more fair. I would probably make a separate cmake option for making native-optimized versions of the benchmark targets that enables stuff like this. Or maybe just compare it to an ISPC based implementation to keep a "generically maintained" version of the serial algorithm that isn't so painfully naïve. Serial is intended to represent how a typical generic implementation would look like but honestly really this whole repo is just about beating whatever the compiler optimizations with non-portable code anyways.

> Different image sizes should be considered? 

Currently checking out a way to have `BigBench` benchmark across several pixel counts up to around 50 mega-pixels since that's currently a pretty fair target for this algorithm since that's around where most flagship smartphones capture at. Also, having the benchmark increment up to that size by around 1 megabyte at a time to make hardware-characteristics like filling an L3 cache more obvious.

> With small image sizes may be good to switch from chrono::high_resolution_clock to `rdtscp`

This is a good point. On ARM I believe `CNTVCT_EL0` would map to that as well.
